Спринт 13. Итерация 1.

ViktorSalk opened the pull request:
Добавил Функциональность «Отзывы».
1) Создал ветку add-reviews
2) Реализовал функции оценки полезности отзывов:
Для новых отзывов она начинается с 0
Для лайков увеличивается на 1
Для дизлайков уменьшается на 1
Отзывы сортируются по степени полезности
3) Реализовал функции получения отзывов по идентификатору фильма, если фильм не указан то все. Если кол-во не указано то 10. 4) Расширил схему SQL
5) Добавил тесты на поведение в оперативной памяти, и в базе данных.

Реализация обеспечивает функциональность для управления отзывами, включая их создание, обновление, удаление и оценку. Постман и прошлые тесты проходит.

Viktor3642 committed:
Спринт 13. Итерация 1.
Добавил Функциональность «Отзывы».
1) Создал ветку add-reviews
2) Реализовал функции оценки полезности отзывов:
Для новых отзывов она начинается с 0
Для лайков увеличивается на 1
Для дизлайков уменьшается на 1
Отзывы сортируются по степени полезности
3) Реализовал функции получения отзывов по идентификатору фильма, если фильм не указан то все. Если кол-во не указано то 10.
4) Расширил схему SQL
5) Добавил тесты на поведение в оперативной памяти, и в базе данных.

Реализация обеспечивает функциональность для управления отзывами, включая их создание, обновление, удаление и оценку.
Постман и прошлые тесты проходит.

Viktor3642 committed:
Спринт 13. Итерация 1.
Добавил Функциональность «Отзывы».
1) Создал ветку add-reviews
2) Реализовал функции оценки полезности отзывов:
Для новых отзывов она начинается с 0
Для лайков увеличивается на 1
Для дизлайков уменьшается на 1
Отзывы сортируются по степени полезности
3) Реализовал функции получения отзывов по идентификатору фильма, если фильм не указан то все. Если кол-во не указано то 10.
4) Расширил схему SQL
5) Добавил тесты на поведение в оперативной памяти, и в базе данных.

Реализация обеспечивает функциональность для управления отзывами, включая их создание, обновление, удаление и оценку.
Постман и прошлые тесты проходит.

polyakovanan commented on src/main/resources/schema.sql:
Точно в рамках этой функциональности нужно вносить изменения в большую часть таблиц?

ViktorSalk replied:
Нет, для функциональности reviews нам нужно только добавить новые таблицы "reviews" и "review_reactions". Существующие таблицы не требуют изменений.

polyakovanan commented on src/main/resources/schema.sql:
```diff
@@ -1,59 +1,65 @@
+    is_positive BOOLEAN NOT NULL,
+    user_id BIGINT NOT NULL REFERENCES users (id) ON DELETE CASCADE,
+    film_id BIGINT NOT NULL REFERENCES films (id) ON DELETE CASCADE,
+    useful INT DEFAULT 0,
```
Это поле как-то будет агрегировать записи из review_reactions?
Ну, то есть, когда добавляется/удаляется запись в review_reactions, то тут значение тоже обновляется?
Да, посмотрел репозиторий, так и есть...
Ощущается немного избыточно, но, зато упрощает маппинг и выборку. Давай так оставим

polyakovanan commented on src/main/java/ru/yandex/practicum/filmorate/storage/dal/repository/ReviewRepository.java:
```diff
@@ -0,0 +1,79 @@
+public class ReviewRepository extends BaseRepository<Review> {
+    private static final String FIND_BY_ID_QUERY = "SELECT * FROM reviews WHERE review_id = ?";
+    private static final String FIND_ALL_QUERY = "SELECT * FROM reviews WHERE film_id = ? OR ? IS NULL ORDER BY useful DESC LIMIT ?";
+    private static final String INSERT_QUERY = "INSERT INTO reviews (content, is_positive, user_id, film_id, useful) VALUES (?, ?, ?, ?, 0)";
```
Вот тут можно не указывать useful, потому что согласно настройкам таблицы там и так будет default 0

ViktorSalk replied:
В ReviewRepository в INSERT_QUERY убрал useful и 0.

polyakovanan commented on src/main/java/ru/yandex/practicum/filmorate/storage/dal/repository/ReviewRepository.java:
```diff
@@ -0,0 +1,79 @@
+    private static final String UPDATE_QUERY = "UPDATE reviews SET content = ?, is_positive = ? WHERE review_id = ?";
+    private static final String DELETE_QUERY = "DELETE FROM reviews WHERE review_id = ?";
+    private static final String ADD_LIKE_QUERY = "INSERT INTO review_reactions (review_id, user_id, is_positive) VALUES (?, ?, true)";
+    private static final String ADD_DISLIKE_QUERY = "INSERT INTO review_reactions (review_id, user_id, is_positive) VALUES (?, ?, false)";
```
Мне кажется, можно объединить ADD_LIKE_QUERY и ADD_DISLIKE_QUERY в один запрос ADD_REACTION_QUERY, куда is_positive добавлять третьим параметром

ViktorSalk replied:
В ReviewRepository объединил ADD_LIKE_QUERY и ADD_DISLIKE_QUERY в ADD_REACTION_QUERY.

polyakovanan commented on src/main/java/ru/yandex/practicum/filmorate/storage/review/DbReviewStorage.java:
```diff
@@ -0,0 +1,71 @@
+    public Review update(Review review) {
+        Optional<Review> existingReview = reviewRepository.findById(review.getReviewId());
+        if (existingReview.isPresent()) {
+            review.setUseful(existingReview.get().getUseful()); // Preserve the useful count
```
Можно условный оператор заменить на лямба-выражение
 existingReview.ifPresent(value -> review.setUseful(value.getUseful()));

ViktorSalk replied:
В DbReviewStorage в update заменил оператор на лямба-выражение.

polyakovanan commented on src/main/java/ru/yandex/practicum/filmorate/storage/dal/repository/ReviewRepository.java:
```diff
@@ -0,0 +1,79 @@
+        update(UPDATE_USEFUL_QUERY, reviewId, reviewId);
+    }
+
+    public void removeDislike(long reviewId, long userId) {
```
Методы removeLike и removeDislike полностью идентичны. Возможно их можно объединить в один

ViktorSalk replied:
В ReviewRepository объединил addLike с addDislike в addReaction и removeLike с removeDislike в removeReaction, обновил зависимости в DbReviewStorage.

polyakovanan commented on src/main/java/ru/yandex/practicum/filmorate/storage/dal/repository/ReviewRepository.java:
```diff
@@ -0,0 +1,79 @@
+        update(UPDATE_USEFUL_QUERY, reviewId, reviewId);
+    }
+
+    public void addDislike(long reviewId, long userId) {
```
Если ADD_LIKE_QUERY и ADD_DISLIKE_QUERY будут объединены по рекомендации выше, то получится та же ситуация, что и у removeLike и removeDislike: методы addLike и addDislike станут полностью идентичны и их можно будет объединить в один

ViktorSalk replied:
В ReviewRepository объединил addLike с addDislike в addReaction и removeLike с removeDislike в removeReaction, обновил зависимости в DbReviewStorage.

polyakovanan commented on src/main/java/ru/yandex/practicum/filmorate/storage/review/InMemoryReviewStorage.java:
```diff
@@ -0,0 +1,98 @@
+                .filter(review -> filmId == null || review.getFilmId().equals(filmId))
+                .sorted(Comparator.comparingInt(Review::getUseful).reversed())
+                .limit(count)
+                .collect(Collectors.toList());
```
Лучше использовать для этого просто метод .toList() вместо collect(). 
toList() обеспечит неизменность получившегося списка, в то время как collect() имеет вероятность косвенно его изменить, например, поменять порядок элементов (в случае, например, многопоточной обработки стрима). Иногда это может быть критично

ViktorSalk replied:
В InMemoryReviewStorage использовал .toList() вместо collect().

polyakovanan commented on src/main/java/ru/yandex/practicum/filmorate/controller/ReviewController.java:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    @DeleteMapping("/{id}/dislike/{userId}")
+    public void removeDislike(@PathVariable Long id, @PathVariable Long userId) {
```
Добавь, пожалуйста, юнит-тесты на удаление лайка/дизлайка. Тогда покрытие тестами будет >90%)

ViktorSalk replied:
Добавил юнит-тесты на удаление лайка/дизлайка в ReviewControllerTest.

Viktor3642 committed:
Merge remote-tracking branch 'origin/add-reviews' into add-reviews

Viktor3642 committed:
Спринт 13. Дорабатываю Функциональность «Отзывы».
1) Добавил юнит-тесты на удаление лайка/дизлайка в ReviewControllerTest.
2) В ReviewRepository в INSERT_QUERY убрал useful и 0.
3) В ReviewRepository объединил ADD_LIKE_QUERY и ADD_DISLIKE_QUERY в ADD_REACTION_QUERY.
4) В ReviewRepository объединил addLike с addDislike в addReaction и removeLike с removeDislike в removeReaction, обновил зависимости в DbReviewStorage.
5) В DbReviewStorage в update заменил оператор на лямба-выражение.
6) В InMemoryReviewStorage использовал .toList() вместо collect().

Viktor3642 committed:
Спринт 13. Дорабатываю Функциональность «Отзывы».
1) Добавил юнит-тесты на удаление лайка/дизлайка в ReviewControllerTest.
2) В ReviewRepository в INSERT_QUERY убрал useful и 0.
3) В ReviewRepository объединил ADD_LIKE_QUERY и ADD_DISLIKE_QUERY в ADD_REACTION_QUERY.
4) В ReviewRepository объединил addLike с addDislike в addReaction и removeLike с removeDislike в removeReaction, обновил зависимости в DbReviewStorage.
5) В DbReviewStorage в update заменил оператор на лямба-выражение.
6) В InMemoryReviewStorage использовал .toList() вместо collect().
7) В InMemoryReviewStorage убрал более неиспользуемый импорт.
